**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: ENMANUEL ABREU
*   MATRÍCULA: a01841472


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [1]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [2]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [25]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [35]:
# Como la columna 'runtime' registra la duración en minutos, filtro las películas cuya duración supera los 300 minutos.

consulta = sqla.text("""
    SELECT *
    FROM movie
    WHERE runtime > 300;
""")

with db.connect() as conn:
    resultado = conn.execute(consulta)
    for fila in resultado:
        print(fila)

(5793, datetime.date(2024, 1, 1), datetime.date(2024, 1, 1), b'\x00', None, '日本統一シリーズ: 映画シリーズ', None, 3892, 'Nihontouitsu Series: Film Series')
(5794, datetime.date(2024, 1, 1), datetime.date(2024, 1, 1), b'\x00', None, '釣りバカ日誌: 映画シリーズ', None, 2120, 'Free and Easy Series: Film Series')
(5874, datetime.date(2024, 1, 1), datetime.date(2024, 1, 1), b'\x01', None, None, datetime.date(2021, 8, 6), 312, 'Navarasa: Limited Series')
(9729, datetime.date(2024, 1, 1), datetime.date(2024, 1, 1), b'\x00', None, '織田同志会 織田征仁: 映画シリーズ', None, 710, 'Seiji Oda: Film Series')
(9730, datetime.date(2024, 1, 1), datetime.date(2024, 1, 1), b'\x00', None, 'キングダム～首領になった男～: 映画シリーズ', None, 427, 'Kingdom ~ The Man Who Became the Top ~: Film Series')


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [36]:
# Liberamos las conexiones previas del pool para evitar saturar el servidor
db.dispose()

# Calculo del porcentaje de películas disponibles únicamente en Estados Unidos (EU)
consulta = sqla.text("""
    SELECT
        ROUND(
            100.0 * SUM(
                CASE
                    WHEN available_globally = FALSE THEN 1
                    ELSE 0
                END
            ) / COUNT(available_globally),
            2
        ) AS porcentaje_peliculas
    FROM movie
    WHERE available_globally IS NOT NULL;
""")
with db.connect() as conn:
    resultado = conn.execute(consulta)
    print(resultado.fetchone())

(Decimal('83.57'),)


**Respuesta:** 83.57%

3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [37]:
# Obtener los diferentes idiomas o regiones originales registrados

consulta = sqla.text("""
    SELECT DISTINCT locale
    FROM movie
    ORDER BY locale;
""")

with db.connect() as conn:
    resultado = conn.execute(consulta)
    for fila in resultado:
        print(fila)

(None,)
('en',)


In [38]:
# Contar cuántas películas tienen el campo 'locale' sin información.

consulta = sqla.text("""
    SELECT COUNT(*) AS cantidad_locale_null
    FROM movie
    WHERE locale IS NULL;
""")

with db.connect() as conn:
    resultado = conn.execute(consulta)
    print(resultado.fetchone())

(11255,)


*   **Idiomas:** en y NULL
*   **Cantidad de peliculas:** 11255

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [39]:
# Contar cuántas películas tienen un título principal diferente
consulta = sqla.text("""
    SELECT COUNT(*) AS titulos_diferentes
    FROM movie
    WHERE title IS NOT NULL
      AND original_title IS NOT NULL
      AND title <> original_title;
""")

with db.connect() as conn:
    resultado = conn.execute(consulta)
    print(resultado.fetchone()[0])

3947


In [40]:
# Comparar la cantidad de películas con locale NULL
consulta = sqla.text("""
    SELECT
        (SELECT COUNT(*)
         FROM movie
         WHERE locale IS NULL) AS cantidad_locale_null,

        (SELECT COUNT(*)
         FROM movie
         WHERE title IS NOT NULL
           AND original_title IS NOT NULL
           AND title <> original_title) AS cantidad_titulos_diferentes;
""")

with db.connect() as conn:
    resultado = conn.execute(consulta)
    print(resultado.fetchone())

(11255, 3947)


In [69]:
# Contar las películas cuyo título principal coincide con su título original.
consulta = sqla.text("""
    SELECT COUNT(*) AS titulos_iguales
    FROM movie
    WHERE title = original_title;
""")


with db.connect() as conn:
    resultado = conn.execute(consulta)
    print(resultado.scalar())

17


*   **Total de peliculas:** 11255
*   **Titulos diferentes:** 3947
*   **Titulos iguales:** 17
*   **Titulos nulos:** 7291

La suposición de que los valores NULL en locale corresponden a películas en un idioma diferente del inglés no es válida. Los 7,291 registros con locale nulo no coinciden con los 3,947 registros que presentan títulos diferentes. Por lo tanto, NULL debe interpretarse como un dato faltante o no especificado en locale, y no como una indicación confiable del idioma de la película.


5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [393]:
# Película con mayor permanencia en el Top 10.
sql_db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)

consulta = sqla.text("""
    SELECT
        m.title,
        MAX(v.cumulative_weeks_in_top10) AS semanas_en_top10
    FROM movie AS m
    INNER JOIN view_summary AS v
        ON m.id = v.movie_id
    WHERE v.duration = 'WEEKLY'
    GROUP BY m.id, m.title
    ORDER BY semanas_en_top10 DESC
    LIMIT 1
""")

with sql_db.connect() as conn:
    resultado = pd.read_sql(consulta, conn)

print(resultado)

           title  semanas_en_top10
0  The Boss Baby                22


**The Boss Baby:** 22 Semanas en el top 10

6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [111]:
# Obtener los 5 programas de TV con mayor cantidad de temporadas.
consulta = sqla.text("""
    SELECT
        tv.title,
        COUNT(s.id) AS cantidad_temporadas
    FROM tv_show AS tv
    INNER JOIN season AS s
        ON tv.id = s.tv_show_id
    GROUP BY tv.id, tv.title
    ORDER BY cantidad_temporadas DESC
    LIMIT 5;
""")

with db.connect() as conn:
    resultado = conn.execute(consulta)

    for fila in resultado:
        print(fila)

("Grey's Anatomy", 20)
('Naruto Shippuden', 20)
('Heartland (2007)', 17)
("It's Always Sunny in Philadelphia", 16)
('NCIS', 15)


1.   **Grey's Anatomy:** 20 temporadas
2.   **Naruto Shippuden:** 20 temporadas
3.   **Heartland:** 17 temporadas
4.   **It's always sunny in Philadelphia:** 16 temporadas
5.   **NCIS:** 15 temporadas

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [316]:
# Obtener los intervalos de fechas de los resúmenes semestrales.
consulta = sqla.text("""
    SELECT DISTINCT
        start_date,
        end_date
    FROM view_summary
    WHERE duration = 'SEMI_ANNUALLY'
    ORDER BY start_date;
""")

with db.connect() as conn:
    resultado = conn.execute(consulta)

    for fila in resultado:
        print(fila)

(datetime.date(2023, 7, 1), datetime.date(2023, 12, 31))
(datetime.date(2024, 1, 1), datetime.date(2024, 6, 30))


**Respuesta:** Los resúmenes semestrales (SEMI_ANNUALLY) corresponden a los períodos del 1 de julio al 31 de diciembre de 2023 y del 1 de enero al 30 de junio de 2024.

8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [314]:
# Obtener las temporadas de Grey's Anatomy y sus vistas
consulta = sqla.text("""
    SELECT
        s.title AS temporada,
        v.views AS cantidad_vistas
    FROM season AS s
    INNER JOIN tv_show AS tv
        ON s.tv_show_id = tv.id
    INNER JOIN view_summary AS v
        ON s.id = v.season_id
    WHERE tv.title = 'Grey''s Anatomy'
      AND v.duration = 'SEMI_ANNUALLY'
      AND v.start_date = '2024-01-01'
      AND v.end_date = '2024-06-30'
    ORDER BY v.views DESC;
""")

with db.connect() as conn:
    resultado = conn.execute(consulta)

    for fila in resultado:
        print(fila)

("Grey's Anatomy: Season 1", 3600000)
("Grey's Anatomy: Season 2", 3100000)
("Grey's Anatomy: Season 3", 2900000)
("Grey's Anatomy: Season 5", 2900000)
("Grey's Anatomy: Season 4", 2900000)
("Grey's Anatomy: Season 6", 2800000)
("Grey's Anatomy: Season 7", 2700000)
("Grey's Anatomy: Season 8", 2600000)
("Grey's Anatomy: Season 9", 2500000)
("Grey's Anatomy: Season 10", 2400000)
("Grey's Anatomy: Season 11", 2200000)
("Grey's Anatomy: Season 12", 2000000)
("Grey's Anatomy: Season 13", 2000000)
("Grey's Anatomy: Season 14", 2000000)
("Grey's Anatomy: Season 19", 2000000)
("Grey's Anatomy: Season 15", 1900000)
("Grey's Anatomy: Season 16", 1800000)
("Grey's Anatomy: Season 18", 1700000)
("Grey's Anatomy: Season 17", 1700000)
("Grey's Anatomy: Season 20", 100000)


**Interpretación:** Las primeras temporadas concentran la mayor cantidad de vistas, mientras que las temporadas posteriores presentan una tendencia a la baja. La temporada 1 supera a la temporada 20 por 3.5 millones de vistas, lo que representa una diferencia muy significativa.

La cantidad de vistas de la temporada 20 podría deberse a que fue estrenada más recientemente y, por tanto, tuvo menos tiempo disponible para acumular vistas durante el período analizado. **No necesariamente significa que sea la temporada menos popular.**

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [352]:
with db.connect() as conn:
    tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
    season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
    view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [353]:
df_unido = season.merge(tv_show, left_on='tv_show_id', right_on='id', suffixes=('_season', '_show'))
df_completo = df_unido.merge(view_summary, left_on='id_season', right_on='season_id')

# Filtrar según las condiciones anteriores
filtro = (
    (df_completo['title_show'] == "Grey's Anatomy") &
    (df_completo['duration'] == 'SEMI_ANNUALLY') &
    (df_completo['start_date'].astype(str) == '2024-01-01') &
    (df_completo['end_date'].astype(str) == '2024-06-30')
)

# Seleccionar, ordenar y mostrar el resultado
resultado_pandas = df_completo[filtro][['title_season', 'views']].sort_values(by='views', ascending=False)
print(resultado_pandas.to_string(index=False))

             title_season   views
 Grey's Anatomy: Season 1 3600000
 Grey's Anatomy: Season 2 3100000
 Grey's Anatomy: Season 3 2900000
 Grey's Anatomy: Season 5 2900000
 Grey's Anatomy: Season 4 2900000
 Grey's Anatomy: Season 6 2800000
 Grey's Anatomy: Season 7 2700000
 Grey's Anatomy: Season 8 2600000
 Grey's Anatomy: Season 9 2500000
Grey's Anatomy: Season 10 2400000
Grey's Anatomy: Season 11 2200000
Grey's Anatomy: Season 12 2000000
Grey's Anatomy: Season 13 2000000
Grey's Anatomy: Season 14 2000000
Grey's Anatomy: Season 19 2000000
Grey's Anatomy: Season 15 1900000
Grey's Anatomy: Season 16 1800000
Grey's Anatomy: Season 18 1700000
Grey's Anatomy: Season 17 1700000
Grey's Anatomy: Season 20  100000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [357]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [360]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [361]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/drive/MyDrive/Ciencia de Datos/Semana 3"
os.chdir(DIR)

In [362]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [394]:
# Obtener la referencia de la colección y recuperar los documentos
empleados_ref = db.collection('EMPLOYEE').stream()

lista_empleados = []
for doc in empleados_ref:
    datos = doc.to_dict()
    datos['id'] = doc.id
    lista_empleados.append(datos)

# Crear el DataFrame
df_empleados = pd.DataFrame(lista_empleados)
print(df_empleados)

  emp_fname   emp_lname                    id              emp_hiredate
0     David      Senior  8LcLuxVHGAd2d9IQc5jF 1989-07-12 06:00:00+00:00
1   William  Smithfield  Fzd60D6Z2CU4n0wVV8YN 2004-06-04 05:00:00+00:00
2      John        News  lX5xuQ5w3i6ib2ExccWY 2000-11-08 06:00:00+00:00
3      June     Arbough  yocFj2lichOkbAj9NBfp 1996-12-01 06:00:00+00:00


In [364]:
firebase_admin.delete_app(firebase_admin.get_app())

Google. (2026). Gemini (basado en Gemini 2.5) [Modelo de lenguaje grande], utilizado para la depuración y optimización del código. https://gemini.google.com/